File: aihound/scanners/persistent_env.py
"""Scanner for OS-level persistent environment variable stores.

Covers:
- Linux/WSL: /etc/environment, /etc/profile.d/*.sh, ~/.pam_environment,
  ~/.config/environment.d/*.conf
- macOS: ~/Library/LaunchAgents/*.plist, /Library/LaunchDaemons/*.plist,
  /etc/launchd.conf
- Windows/WSL: HKCU\\Environment and HKLM\\...\\Environment via reg.exe
"""

from __future__ import annotations

import logging
import re
import subprocess
from pathlib import Path
from typing import Optional

logger = logging.getLogger("aihound.scanners.persistent_env")

from aihound.core.scanner import (
    BaseScanner,
    CredentialFinding,
    ScanResult,
    StorageType,
    RiskLevel,
)
from aihound.core.platform import detect_platform, Platform, get_home, get_wsl_windows_home
from aihound.core.redactor import mask_value, identify_credential_type, KNOWN_PREFIXES
from aihound.core.permissions import (
    get_file_permissions,
    get_file_owner,
    get_file_mtime,
    describe_staleness,
)
from aihound.remediation import hint_manual, hint_run_command
from aihound.scanners import register
from aihound.scanners.envvars import AI_ENV_VARS

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_MAX_PROFILE_D_SIZE = 65536  # 64KB

# ---------------------------------------------------------------------------
# Compiled regex patterns
# ---------------------------------------------------------------------------

# /etc/environment and environment.d: VAR=value (no "export" prefix)
_KV_RE = re.compile(
    r"""(?mx)^ \s* ([A-Z_][A-Z0-9_]*) \s* = \s* ["']? ([^"'\s#]+) ["']?"""
)

# profile.d / launchd.conf: export VAR=value
_EXPORT_RE = re.compile(
    r"""(?mx)^ \s* export \s+ ([A-Z_][A-Z0-9_]*) \s* = \s* ["']? ([^"'\s#]+) ["']?"""
)

# ~/.pam_environment: VAR DEFAULT=value  or  VAR OVERRIDE=value
_PAM_RE = re.compile(
    r"""(?mx)^ \s* ([A-Z_][A-Z0-9_]*) \s+ (?:DEFAULT|OVERRIDE) \s* = \s* ["']? ([^"'\s#]+) ["']?"""
)

# reg.exe query output:    NAME    REG_SZ    VALUE
_REG_LINE_RE = re.compile(
    r"^\s+(\S+)\s+REG_(?:SZ|EXPAND_SZ)\s+(.+)$", re.MULTILINE
)


# ---------------------------------------------------------------------------
# Helper
# ---------------------------------------------------------------------------

def _is_ai_relevant(var_name: str, value: str) -> bool:
    """Return True if var_name is known AI cred or value matches a known prefix."""
    if var_name in AI_ENV_VARS:
        return True
    if identify_credential_type(value):
        return True
    return False


# ---------------------------------------------------------------------------
# Scanner
# ---------------------------------------------------------------------------

@register
class PersistentEnvScanner(BaseScanner):
    """Scan OS-level persistent environment variable stores."""

    def name(self) -> str:
        return "Persistent Environment"

    def slug(self) -> str:
        return "persistent-env"

    # ------------------------------------------------------------------
    # Public scan entry point
    # ------------------------------------------------------------------

    def scan(self, show_secrets: bool = False) -> ScanResult:
        plat = detect_platform()
        result = ScanResult(scanner_name=self.name(), platform=plat.value)

        if plat in (Platform.LINUX, Platform.WSL):
            self._scan_linux(result, show_secrets)
        elif plat == Platform.MACOS:
            self._scan_macos(result, show_secrets)
        elif plat == Platform.WINDOWS:
            self._scan_windows_registry(result, show_secrets)

        # On WSL also check the Windows registry
        if plat == Platform.WSL:
            self._scan_windows_registry(result, show_secrets)

        return result

    # ------------------------------------------------------------------
    # Linux / WSL scanning
    # ------------------------------------------------------------------

    def _scan_linux(self, result: ScanResult, show_secrets: bool) -> None:
        # /etc/environment — system-wide key=value
        etc_env = Path("/etc/environment")
        text = self._read_file(etc_env, result)
        if text is not None:
            findings = self._scan_kv_content(text, etc_env, show_secrets, is_system=True)
            result.findings.extend(findings)

        # /etc/profile.d/*.sh — system-wide shell scripts
        profile_d = Path("/etc/profile.d")
        if profile_d.is_dir():
            for sh_file in sorted(profile_d.glob("*.sh")):
                try:
                    size = sh_file.stat().st_size
                except OSError:
                    continue
                if size > _MAX_PROFILE_D_SIZE:
                    logger.debug("Skipping large profile.d file: %s (%d bytes)", sh_file, size)
                    result.errors.append(f"Skipped (>64KB): {sh_file}")
                    continue
                text = self._read_file(sh_file, result)
                if text is not None:
                    findings = self._scan_export_content(text, sh_file, show_secrets, is_system=True)
                    result.findings.extend(findings)

        # ~/.pam_environment — user-level PAM
        pam_env = get_home() / ".pam_environment"
        text = self._read_file(pam_env, result)
        if text is not None:
            findings = self._scan_pam_content(text, pam_env, show_secrets)
            result.findings.extend(findings)

        # ~/.config/environment.d/*.conf — systemd user env
        env_d = get_home() / ".config" / "environment.d"
        if env_d.is_dir():
            for conf_file in sorted(env_d.glob("*.conf")):
                text = self._read_file(conf_file, result)
                if text is not None:
                    findings = self._scan_kv_content(text, conf_file, show_secrets, is_system=False)
                    result.findings.extend(findings)

    # ------------------------------------------------------------------
    # macOS scanning
    # ------------------------------------------------------------------

    def _scan_macos(self, result: ScanResult, show_secrets: bool) -> None:
        import plistlib

        # ~/Library/LaunchAgents/*.plist — user-level
        user_agents = get_home() / "Library" / "LaunchAgents"
        if user_agents.is_dir():
            for plist_file in sorted(user_agents.glob("*.plist")):
                self._scan_plist_file(plist_file, result, show_secrets, is_system=False)

        # /Library/LaunchDaemons/*.plist — system-level
        sys_daemons = Path("/Library/LaunchDaemons")
        if sys_daemons.is_dir():
            for plist_file in sorted(sys_daemons.glob("*.plist")):
                self._scan_plist_file(plist_file, result, show_secrets, is_system=True)

        # /etc/launchd.conf — deprecated export-style
        launchd_conf = Path("/etc/launchd.conf")
        text = self._read_file(launchd_conf, result)
        if text is not None:
            findings = self._scan_export_content(text, launchd_conf, show_secrets, is_system=True)
            result.findings.extend(findings)

    def _scan_plist_file(
        self, path: Path, result: ScanResult, show_secrets: bool, is_system: bool
    ) -> None:
        import plistlib

        try:
            with open(path, "rb") as fh:
                data = plistlib.load(fh)
        except Exception as exc:
            logger.debug("Could not parse plist %s: %s", path, exc)
            return

        if not isinstance(data, dict):
            return

        env_dict = data.get("EnvironmentVariables")
        if not isinstance(env_dict, dict):
            return

        findings = self._scan_plist_env_dict(env_dict, path, show_secrets, is_system=is_system)
        result.findings.extend(findings)

    # ------------------------------------------------------------------
    # Windows registry scanning
    # ------------------------------------------------------------------

    def _scan_windows_registry(self, result: ScanResult, show_secrets: bool) -> None:
        hkcu_key = r"HKCU\Environment"
        hklm_key = r"HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment"

        output = self._query_registry(hkcu_key, result)
        if output is not None:
            findings = self._parse_reg_output(output, hkcu_key, show_secrets, is_system=False)
            result.findings.extend(findings)

        output = self._query_registry(hklm_key, result)
        if output is not None:
            findings = self._parse_reg_output(output, hklm_key, show_secrets, is_system=True)
            result.findings.extend(findings)

    # ------------------------------------------------------------------
    # Content parsers
    # ------------------------------------------------------------------

    def _scan_kv_content(
        self,
        text: str,
        path: Path,
        show_secrets: bool,
        is_system: bool,
    ) -> list[CredentialFinding]:
        findings: list[CredentialFinding] = []
        for line_num, match in self._matches_with_line_numbers(text, _KV_RE):
            var_name, value = match.group(1), match.group(2).strip()
            if not _is_ai_relevant(var_name, value):
                continue
            findings.append(
                self._make_file_finding(var_name, value, path, line_num, show_secrets, is_system)
            )
        return findings

    def _scan_export_content(
        self,
        text: str,
        path: Path,
        show_secrets: bool,
        is_system: bool,
    ) -> list[CredentialFinding]:
        findings: list[CredentialFinding] = []
        for line_num, match in self._matches_with_line_numbers(text, _EXPORT_RE):
            var_name, value = match.group(1), match.group(2).strip()
            if not _is_ai_relevant(var_name, value):
                continue
            findings.append(
                self._make_file_finding(var_name, value, path, line_num, show_secrets, is_system)
            )
        return findings

    def _scan_pam_content(
        self,
        text: str,
        path: Path,
        show_secrets: bool,
    ) -> list[CredentialFinding]:
        findings: list[CredentialFinding] = []
        for line_num, match in self._matches_with_line_numbers(text, _PAM_RE):
            var_name, value = match.group(1), match.group(2).strip()
            if not _is_ai_relevant(var_name, value):
                continue
            # ~/.pam_environment is always user-level (HIGH)
            findings.append(
                self._make_file_finding(
                    var_name,
                    value,
                    path,
                    line_num,
                    show_secrets,
                    is_system=False,
                    storage_override=StorageType.PLAINTEXT_ENV,
                )
            )
        return findings

    def _scan_plist_env_dict(
        self,
        env_dict: dict,
        path: Path,
        show_secrets: bool,
        is_system: bool,
    ) -> list[CredentialFinding]:
        findings: list[CredentialFinding] = []
        for var_name, value in env_dict.items():
            if not isinstance(value, str):
                continue
            if not _is_ai_relevant(var_name, value):
                continue

            cred_type = identify_credential_type(value) or AI_ENV_VARS.get(var_name, var_name)
            risk = RiskLevel.CRITICAL if is_system else RiskLevel.HIGH
            notes = [f"Found {var_name} in EnvironmentVariables dict of {path.name}"]

            findings.append(
                CredentialFinding(
                    tool_name=self.name(),
                    credential_type=cred_type,
                    storage_type=StorageType.PLAINTEXT_FILE,
                    location=str(path),
                    exists=True,
                    risk_level=risk,
                    value_preview=mask_value(value, show_full=show_secrets),
                    raw_value=value if show_secrets else None,
                    file_permissions=get_file_permissions(path),
                    file_owner=get_file_owner(path),
                    file_modified=get_file_mtime(path),
                    notes=notes,
                    remediation="Remove credential from persistent environment store and use a secret manager",
                    remediation_hint=hint_manual(
                        "Remove credential from persistent environment store and use a secret manager",
                        suggested_tools=["1Password CLI", "doppler", "vault"],
                    ),
                )
            )
        return findings

    def _parse_reg_output(
        self,
        output: str,
        reg_key: str,
        show_secrets: bool,
        is_system: bool,
    ) -> list[CredentialFinding]:
        findings: list[CredentialFinding] = []
        for match in _REG_LINE_RE.finditer(output):
            var_name = match.group(1).strip()
            value = match.group(2).strip()
            if not _is_ai_relevant(var_name, value):
                continue

            cred_type = identify_credential_type(value) or AI_ENV_VARS.get(var_name, var_name)
            risk = RiskLevel.CRITICAL if is_system else RiskLevel.HIGH
            scope = "Machine" if is_system else "User"
            ps_cmd = f"[System.Environment]::SetEnvironmentVariable('{var_name}', $null, '{scope}')"

            findings.append(
                CredentialFinding(
                    tool_name=self.name(),
                    credential_type=cred_type,
                    storage_type=StorageType.PLAINTEXT_INI,
                    location=f"{reg_key}\\{var_name}",
                    exists=True,
                    risk_level=risk,
                    value_preview=mask_value(value, show_full=show_secrets),
                    raw_value=value if show_secrets else None,
                    notes=[f"Registry key: {reg_key}"],
                    remediation="Remove credential from Windows registry environment and use a secret manager",
                    remediation_hint=hint_run_command(
                        [ps_cmd],
                        shell="powershell",
                    ),
                )
            )
        return findings

    # ------------------------------------------------------------------
    # Low-level helpers
    # ------------------------------------------------------------------

    def _read_file(self, path: Path, result: ScanResult) -> Optional[str]:
        """Read a file, returning its text or None (adding to errors on failure)."""
        if not path.exists():
            return None
        try:
            return path.read_text(encoding="utf-8", errors="replace")
        except PermissionError:
            result.errors.append(f"Permission denied: {path}")
        except OSError as exc:
            result.errors.append(f"Could not read {path}: {exc}")
        return None

    def _query_registry(self, key: str, result: ScanResult) -> Optional[str]:
        """Run reg.exe query and return stdout, or None on failure."""
        try:
            proc = subprocess.run(
                ["reg.exe", "query", key],
                capture_output=True,
                text=True,
                timeout=5,
            )
            return proc.stdout
        except FileNotFoundError:
            # reg.exe not available (pure Linux)
            return None
        except subprocess.TimeoutExpired:
            result.errors.append(f"Timeout querying registry key: {key}")
            return None
        except OSError as exc:
            result.errors.append(f"OSError querying registry key {key}: {exc}")
            return None

    def _make_file_finding(
        self,
        var_name: str,
        value: str,
        path: Path,
        line_num: int,
        show_secrets: bool,
        is_system: bool,
        storage_override: Optional[StorageType] = None,
    ) -> CredentialFinding:
        cred_type = identify_credential_type(value) or AI_ENV_VARS.get(var_name, var_name)
        risk = RiskLevel.CRITICAL if is_system else RiskLevel.HIGH
        storage = storage_override if storage_override is not None else StorageType.PLAINTEXT_FILE
        notes = [f"Found {var_name} at line {line_num} in {path}"]

        return CredentialFinding(
            tool_name=self.name(),
            credential_type=cred_type,
            storage_type=storage,
            location=f"{path}:{line_num}",
            exists=True,
            risk_level=risk,
            value_preview=mask_value(value, show_full=show_secrets),
            raw_value=value if show_secrets else None,
            file_permissions=get_file_permissions(path),
            file_owner=get_file_owner(path),
            file_modified=get_file_mtime(path),
            notes=notes,
            remediation="Remove credential from persistent environment store and use a secret manager",
            remediation_hint=hint_manual(
                "Remove credential from persistent environment store and use a secret manager",
                suggested_tools=["1Password CLI", "doppler", "vault"],
            ),
        )

    @staticmethod
    def _matches_with_line_numbers(
        text: str, pattern: re.Pattern
    ) -> list[tuple[int, re.Match]]:
        """Return (1-based line number, match) pairs for all regex matches in text."""
        results = []
        for match in pattern.finditer(text):
            line_num = text.count("\n", 0, match.start()) + 1
            results.append((line_num, match))
        return results

File: Other Versions/Go/scanners/persistent_env.go
package scanners

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"aihound/core"
	"aihound/remediation"
)

const maxProfileDSize = 65536 // 64KB

// Compiled regex patterns for persistent environment scanning.
var (
	// /etc/environment and environment.d: VAR=value (no "export" prefix)
	peKVRe = regexp.MustCompile(`(?m)^\s*([A-Z_][A-Z0-9_]*)\s*=\s*["']?([^"'\s#]+)["']?`)

	// profile.d / launchd.conf: export VAR=value
	peExportRe = regexp.MustCompile(`(?m)^\s*export\s+([A-Z_][A-Z0-9_]*)\s*=\s*["']?([^"'\s#]+)["']?`)

	// ~/.pam_environment: VAR DEFAULT=value  or  VAR OVERRIDE=value
	pePamRe = regexp.MustCompile(`(?m)^\s*([A-Z_][A-Z0-9_]*)\s+(?:DEFAULT|OVERRIDE)\s*=\s*["']?([^"'\s#]+)["']?`)

	// reg.exe query output:    NAME    REG_SZ    VALUE
	peRegLineRe = regexp.MustCompile(`(?m)^\s+(\S+)\s+REG_(?:SZ|EXPAND_SZ)\s+(.+)$`)
)

// peIsAIRelevant returns true if the variable name is a known AI env var
// or the value matches a known credential prefix.
func peIsAIRelevant(varName, value string) bool {
	if _, ok := AIEnvVars[varName]; ok {
		return true
	}
	if core.IdentifyCredentialType(value) != "" {
		return true
	}
	return false
}

type persistentEnvScanner struct{}

func init() {
	Register(&persistentEnvScanner{})
}

func (s *persistentEnvScanner) Name() string      { return "Persistent Environment" }
func (s *persistentEnvScanner) Slug() string       { return "persistent-env" }
func (s *persistentEnvScanner) IsApplicable() bool { return true }

func (s *persistentEnvScanner) Scan(showSecrets bool) core.ScanResult {
	plat := core.DetectPlatform()
	result := core.ScanResult{ScannerName: s.Name(), Platform: plat.String()}

	if plat == core.PlatformLinux || plat == core.PlatformWSL {
		s.scanLinux(&result, showSecrets)
	} else if plat == core.PlatformMacOS {
		s.scanMacOS(&result, showSecrets)
	} else if plat == core.PlatformWindows {
		s.scanWindowsRegistry(&result, showSecrets)
	}

	// On WSL also check the Windows registry
	if plat == core.PlatformWSL {
		s.scanWindowsRegistry(&result, showSecrets)
	}

	return result
}

// ---------------------------------------------------------------------------
// Linux / WSL scanning
// ---------------------------------------------------------------------------

func (s *persistentEnvScanner) scanLinux(result *core.ScanResult, showSecrets bool) {
	// /etc/environment -- system-wide key=value
	etcEnv := "/etc/environment"
	if text, ok := s.readFile(etcEnv, result); ok {
		for _, f := range s.scanKVContent(text, etcEnv, showSecrets, true) {
			result.Findings = append(result.Findings, f)
		}
	}

	// /etc/profile.d/*.sh -- system-wide shell scripts
	profileD := "/etc/profile.d"
	if info, err := os.Stat(profileD); err == nil && info.IsDir() {
		matches, _ := filepath.Glob(filepath.Join(profileD, "*.sh"))
		for _, shFile := range matches {
			fi, err := os.Stat(shFile)
			if err != nil {
				continue
			}
			if fi.Size() > maxProfileDSize {
				result.Errors = append(result.Errors, fmt.Sprintf("Skipped (>64KB): %s", shFile))
				continue
			}
			if text, ok := s.readFile(shFile, result); ok {
				for _, f := range s.scanExportContent(text, shFile, showSecrets, true) {
					result.Findings = append(result.Findings, f)
				}
			}
		}
	}

	// ~/.pam_environment -- user-level PAM
	home := core.GetHome()
	pamEnv := filepath.Join(home, ".pam_environment")
	if text, ok := s.readFile(pamEnv, result); ok {
		for _, f := range s.scanPamContent(text, pamEnv, showSecrets) {
			result.Findings = append(result.Findings, f)
		}
	}

	// ~/.config/environment.d/*.conf -- systemd user env
	envD := filepath.Join(home, ".config", "environment.d")
	if info, err := os.Stat(envD); err == nil && info.IsDir() {
		matches, _ := filepath.Glob(filepath.Join(envD, "*.conf"))
		for _, confFile := range matches {
			if text, ok := s.readFile(confFile, result); ok {
				for _, f := range s.scanKVContent(text, confFile, showSecrets, false) {
					result.Findings = append(result.Findings, f)
				}
			}
		}
	}
}

// ---------------------------------------------------------------------------
// macOS scanning
// ---------------------------------------------------------------------------

func (s *persistentEnvScanner) scanMacOS(result *core.ScanResult, showSecrets bool) {
	// TODO: macOS plist parsing (~/Library/LaunchAgents/*.plist,
	// /Library/LaunchDaemons/*.plist) is skipped because Go's stdlib
	// has no plistlib equivalent. Consider howett.net/plist if needed.

	// /etc/launchd.conf -- deprecated export-style
	launchdConf := "/etc/launchd.conf"
	if text, ok := s.readFile(launchdConf, result); ok {
		for _, f := range s.scanExportContent(text, launchdConf, showSecrets, true) {
			result.Findings = append(result.Findings, f)
		}
	}
}

// ---------------------------------------------------------------------------
// Windows registry scanning
// ---------------------------------------------------------------------------

func (s *persistentEnvScanner) scanWindowsRegistry(result *core.ScanResult, showSecrets bool) {
	hkcuKey := `HKCU\Environment`
	hklmKey := `HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment`

	if output, ok := s.queryRegistry(hkcuKey, result); ok {
		for _, f := range s.parseRegOutput(output, hkcuKey, showSecrets, false) {
			result.Findings = append(result.Findings, f)
		}
	}

	if output, ok := s.queryRegistry(hklmKey, result); ok {
		for _, f := range s.parseRegOutput(output, hklmKey, showSecrets, true) {
			result.Findings = append(result.Findings, f)
		}
	}
}

// ---------------------------------------------------------------------------
// Content parsers
// ---------------------------------------------------------------------------

func (s *persistentEnvScanner) scanKVContent(text, path string, showSecrets, isSystem bool) []core.CredentialFinding {
	var findings []core.CredentialFinding
	for _, pair := range peMatchesWithLineNumbers(text, peKVRe) {
		varName := pair.match[1]
		value := strings.TrimSpace(pair.match[2])
		if !peIsAIRelevant(varName, value) {
			continue
		}
		findings = append(findings, s.makeFileFinding(varName, value, path, pair.lineNum, showSecrets, isSystem, core.PlaintextFile))
	}
	return findings
}

func (s *persistentEnvScanner) scanExportContent(text, path string, showSecrets, isSystem bool) []core.CredentialFinding {
	var findings []core.CredentialFinding
	for _, pair := range peMatchesWithLineNumbers(text, peExportRe) {
		varName := pair.match[1]
		value := strings.TrimSpace(pair.match[2])
		if !peIsAIRelevant(varName, value) {
			continue
		}
		findings = append(findings, s.makeFileFinding(varName, value, path, pair.lineNum, showSecrets, isSystem, core.PlaintextFile))
	}
	return findings
}

func (s *persistentEnvScanner) scanPamContent(text, path string, showSecrets bool) []core.CredentialFinding {
	var findings []core.CredentialFinding
	for _, pair := range peMatchesWithLineNumbers(text, pePamRe) {
		varName := pair.match[1]
		value := strings.TrimSpace(pair.match[2])
		if !peIsAIRelevant(varName, value) {
			continue
		}
		// ~/.pam_environment is always user-level (HIGH via PlaintextENV)
		findings = append(findings, s.makeFileFinding(varName, value, path, pair.lineNum, showSecrets, false, core.PlaintextENV))
	}
	return findings
}

func (s *persistentEnvScanner) parseRegOutput(output, regKey string, showSecrets, isSystem bool) []core.CredentialFinding {
	var findings []core.CredentialFinding
	for _, m := range peRegLineRe.FindAllStringSubmatch(output, -1) {
		varName := strings.TrimSpace(m[1])
		value := strings.TrimSpace(m[2])
		if !peIsAIRelevant(varName, value) {
			continue
		}

		credType := core.IdentifyCredentialType(value)
		if credType == "" {
			if desc, ok := AIEnvVars[varName]; ok {
				credType = desc
			} else {
				credType = varName
			}
		}

		risk := core.RiskHigh
		if isSystem {
			risk = core.RiskCritical
		}

		scope := "User"
		if isSystem {
			scope = "Machine"
		}
		psCmd := fmt.Sprintf("[System.Environment]::SetEnvironmentVariable('%s', $null, '%s')", varName, scope)

		rawValue := ""
		if showSecrets {
			rawValue = value
		}

		findings = append(findings, core.CredentialFinding{
			ToolName:       s.Name(),
			CredentialType: credType,
			StorageType:    core.PlaintextINI,
			Location:       fmt.Sprintf(`%s\%s`, regKey, varName),
			Exists:         true,
			RiskLevel:      risk,
			ValuePreview:   core.MaskValue(value, showSecrets),
			RawValue:       rawValue,
			Notes:          []string{fmt.Sprintf("Registry key: %s", regKey)},
			Remediation:    "Remove credential from Windows registry environment and use a secret manager",
			RemediationHint: remediation.HintRunCommand(
				[]string{psCmd},
				"powershell",
			),
		})
	}
	return findings
}

// ---------------------------------------------------------------------------
// Low-level helpers
// ---------------------------------------------------------------------------

func (s *persistentEnvScanner) readFile(path string, result *core.ScanResult) (string, bool) {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return "", false
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsPermission(err) {
			result.Errors = append(result.Errors, fmt.Sprintf("Permission denied: %s", path))
		} else {
			result.Errors = append(result.Errors, fmt.Sprintf("Could not read %s: %s", path, err))
		}
		return "", false
	}
	return string(data), true
}

func (s *persistentEnvScanner) queryRegistry(key string, result *core.ScanResult) (string, bool) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "reg.exe", "query", key)
	out, err := cmd.Output()
	if err != nil {
		// reg.exe not available (pure Linux) or key doesn't exist -- not an error
		if ctx.Err() == context.DeadlineExceeded {
			result.Errors = append(result.Errors, fmt.Sprintf("Timeout querying registry key: %s", key))
		}
		return "", false
	}
	return string(out), true
}

func (s *persistentEnvScanner) makeFileFinding(
	varName, value, path string, lineNum int, showSecrets, isSystem bool, storage core.StorageType,
) core.CredentialFinding {
	credType := core.IdentifyCredentialType(value)
	if credType == "" {
		if desc, ok := AIEnvVars[varName]; ok {
			credType = desc
		} else {
			credType = varName
		}
	}

	risk := core.RiskHigh
	if isSystem {
		risk = core.RiskCritical
	}

	rawValue := ""
	if showSecrets {
		rawValue = value
	}

	return core.CredentialFinding{
		ToolName:        s.Name(),
		CredentialType:  credType,
		StorageType:     storage,
		Location:        fmt.Sprintf("%s:%d", path, lineNum),
		Exists:          true,
		RiskLevel:       risk,
		ValuePreview:    core.MaskValue(value, showSecrets),
		RawValue:        rawValue,
		FilePermissions: core.GetFilePermissions(path),
		FileOwner:       core.GetFileOwner(path),
		FileModified:    core.GetFileMtime(path),
		Notes:           []string{fmt.Sprintf("Found %s at line %d in %s", varName, lineNum, path)},
		Remediation:     "Remove credential from persistent environment store and use a secret manager",
		RemediationHint: remediation.HintManual(
			"Remove credential from persistent environment store and use a secret manager",
			map[string]any{"suggested_tools": []string{"1Password CLI", "doppler", "vault"}},
		),
	}
}

// peLineMatch holds a regex submatch and its 1-based line number.
type peLineMatch struct {
	lineNum int
	match   []string // submatch groups
}

// peMatchesWithLineNumbers returns all matches with their 1-based line numbers.
func peMatchesWithLineNumbers(text string, pattern *regexp.Regexp) []peLineMatch {
	var results []peLineMatch
	for _, loc := range pattern.FindAllStringSubmatchIndex(text, -1) {
		lineNum := strings.Count(text[:loc[0]], "\n") + 1
		// Extract submatch strings
		groups := make([]string, 0, len(loc)/2)
		for i := 0; i < len(loc); i += 2 {
			if loc[i] >= 0 {
				groups = append(groups, text[loc[i]:loc[i+1]])
			} else {
				groups = append(groups, "")
			}
		}
		results = append(results, peLineMatch{lineNum: lineNum, match: groups})
	}
	return results
}
